firedrake/mg/opencascade_mh.py

```diff
@@ -0,0 +1,196 @@
+    coorddm = plex.getCoordinateDM()
+    x = mesh.coordinates
+    nodesection = x.function_space().dm.getDefaultSection()
+
+    dofsection = PETSc.Section().create(comm=mesh.mpi_comm())
+    dofsection.setNumFields(1)
+    dofsection.setFieldComponents(0, gdim)
+    (lo, hi) = nodesection.getChart()
+    dofsection.setChart(lo, hi)
+    for point in range(lo, hi):
+        dof = nodesection.getDof(point)
+        dofsection.setDof(point, dof*gdim)
+        off = nodesection.getOffset(point)
+        dofsection.setOffset(point, off*gdim)
+    coorddm.setDefaultSection(dofsection)
+
+    with x.dat.vec_ro as x_:
+        plex.setCoordinates(x_)
+        plex.setCoordinatesLocal(x_)
+
+
+def project_mesh_to_cad(mesh, cad):
+
+    from OCC.Core.BRepAdaptor import BRepAdaptor_Surface, BRepAdaptor_Curve
+    from OCC.Core.gp import gp_Pnt
+    from OCC.Core.GeomAPI import GeomAPI_ProjectPointOnSurf, GeomAPI_ProjectPointOnCurve
+
+    coorddata = mesh.coordinates.dat.data
+    ids = mesh.exterior_facets.unique_markers
+
+    filt = lambda arr: arr[numpy.where(arr < mesh.coordinates.dof_dset.size)[0]]
+    boundary_nodes = {id: filt(mesh.coordinates.function_space().boundary_nodes(int(id), "topological")) for id in ids}
+
+    for (id, face) in zip(ids, cad.faces()):
+        owned_nodes = boundary_nodes[id]
+        for other_id in ids:
+            if id == other_id:
+                continue
+            owned_nodes = numpy.setdiff1d(owned_nodes, boundary_nodes[other_id])
+
+        surf = BRepAdaptor_Surface(face)
+
+        for node in owned_nodes:
+            pt = gp_Pnt(*coorddata[node, :])
+
+            proj = GeomAPI_ProjectPointOnSurf(pt, surf.Surface().Surface())
+            if proj.NbPoints() > 0:
+                projpt = proj.NearestPoint()
+                coorddata[node, :] = projpt.Coord()
+            else:
+                warnings.warn("Projection of point %s onto face %d failed" % (coordata[node, :], id))
+
+        edges = set(cad.edges_from_face(face))
+
+        for (other_id, other_face) in zip(ids, cad.faces()):
+            if other_id <= id:
+                continue
+
+            intersecting_nodes = numpy.intersect1d(boundary_nodes[id], boundary_nodes[other_id])
+            if len(intersecting_nodes) == 0:
+                continue
+
+            other_edges = set(cad.edges_from_face(other_face))
+
+            intersecting_edges = []
+            for edge in edges:
+                s = str(edge)  # FIXME: is there a more elegant way to get the OCC id?
+                for other_edge in other_edges:
+                    other_s = str(other_edge)
+                    if s == other_s:
+                        intersecting_edges.append(edge)
+
+            if len(intersecting_edges) == 0:
+                warnings.warn("face: %s other_face: %s intersecting_edges: %s" % (face, other_face, intersecting_edges))
+                warnings.warn("Warning: no intersecting edges in CAD, even though vertices on both faces?")
```

Comment:
I tried to think of a case where this might happen, but only in 4D?

Reply:
I've seen it once with CAD data that wasn't well-formed.